### ISSUE TITLE
UEFI support for KVM

### ISSUE THREAD

**Issue by jamesarems** [OpenNebula/one#2481]
Hi,
Do opennebula have UEFI support? 
We tried to change bootloader and custom templates but no luck.
I have tried to boot iso with uefi using virt-install and its working perfectly.

Please suggest the workaround , because we didn't find any documentation regarding uefi on your docs.

**Comment by cloudweavers** [comment]
Given the fact that some VM now require UEFI to boot, I would say it can be an important addition. OVMF is widely available on nearly all linux distributions, and on the ONE side it would require just the addition of a "BIOS" section.

**Comment by mzealey** [comment]
A simple thing is to do:

```
    <os><loader readonly='yes' type='pflash'>/usr/share/OVMF/OVMF_CODE.secboot.fd</loader></os>
```

But this can have issues due to the default nvram location being `/var/lib/libvirt/qemu/nvram` which is not accessible by opennebula user oneadmin by default. It would be better to store the nvram with the disk images in the datastore directory as with the context disk. This can be done with the `<nvram>` option.

**Comment by altaic** [comment]
Any progress on adding UEFI support? Or otherwise a work-around?

**Comment by GnunuX** [comment]
I've just open an issue on libvirt project: https://gitlab.com/libvirt/libvirt/-/issues/159
With my modification, i just need to add:

```RAW = [
  DATA = "
  <os firmware='efi'>
    <type>hvm</type>
    <loader secure='no'/>
  </os>
",
  TYPE = "KVM" ]```

to template configuration and VM starts with EFI support.

**Comment by josephtingiris** [comment]
There's a workaround for this that doesn't require waiting on upstream changes to libvirtd.  Mount /var/lib/libvirt/qemu/nvram with a filesystem that supports setting uid=oneadmin.  libvirtd doesn't know the difference & UEFI VMs, including migrations, work fine.

There are multiple filesystems that support uid= (see man mount).  Below is how I do it, using a loopback device, autofs, & vfat.   If you don't want to use autofs, /etc/fstab works.   Loopback devices aren't required, either.  Any device that can be formatted & mounted with uid=oneadmin will work.

```
cd /var/lib/libvirt/qemu
dd if=/dev/zero of=/var/lib/libvirt/qemu/nvram.img bs=100M count=2 # 200M
chown root. /var/lib/libvirt/qemu/nvram.img && chmod 0600 /var/lib/libvirt/qemu/nvram.img # only root needs to rw the image
losetup -fP /var/lib/libvirt/qemu/nvram.img && losetup -a # add & print loopback devices
mkfs.vfat /dev/loop0 # format image with vfat
losetup -d /dev/loop0 # remove loopback device

# install autofs; setup direct automount map
chown oneadmin. /var/lib/libvirt/qemu/nvram # autofs will change the mode via umask
echo "/- /etc/auto.qemu-nvram" > /etc/auto.master.d/nvram.autofs
echo "/var/lib/libvirt/qemu/nvram -fstype=vfat,uid=oneadmin,gid=oneadmin,umask=0077 :/var/lib/libvirt/qemu/nvram.img" > /etc/auto.qemu-nvram
systemctl restart autofs

touch /var/lib/libvirt/qemu/nvram/test && ls -l /var/lib/libvirt/qemu/nvram && rm /var/lib/libvirt/qemu/nvram/test
```

**Comment by FrederickBor** [comment]
@tinova PRs to be approved:
- [Code](https://github.com/OpenNebula/one-ee/pull/1452)
- [Docs](https://github.com/OpenNebula/docs/pull/1967)

**Comment by atodorov-storpool** [comment]
Having some experience with UEFI boot with OIpenNebula and after looking at the code I would like to understand some of the insights backing the current implementation?

* There is no option/definition of the `secure` attribute set for the `loader` element at https://github.com/OpenNebula/one/commit/121b4b096ee36d3e1756b868a79e7d03dfb8b13d#diff-4f32f4a94d596188452be83043bf54091d4ded4e0d9e4b8fde332e124a3fa9fbR807 . It is an assumption that all distributions had qemu-kvm builds with secure=on enabled by default?

* There is no option to define the [SMM](https://libvirt.org/formatdomain.html#hypervisor-features) feature (SMM is needed for the Secure Boot). So it is assumed that it is active and enabled by default?

* Also, the NVRAM is a container for variables stored by the VM OS. For example, it is used as a secure key-value store by Windows (enterprise) software like boot order, programs activation, licenses, etc. That said (IMHO) the backing entity for the NVRAM container should be handled by the storage drivers because it could be a plain file or a block device and considered as (sort of) an additional "disk" that must be included in the entire VM lifecycle, including backups, snapshots, etc.

What do you think?

**Comment by hydro-b** [comment]
>     * Also, the NVRAM is a container for variables stored by the VM OS. For example, it is used as a secure key-value store by Windows (enterprise) software like boot order, programs activation, licenses, etc. That said (IMHO) the backing entity for the NVRAM container should be handled by the storage drivers because it could be a plain file or a block device and considered as (sort of) an additional "disk" that must be included in the entire VM lifecycle, including backups, snapshots, etc.

I  agree, I too would like to see this handled by the storage drivers.

**Comment by rsmontero** [comment]
About storage, it is good idea to be handled by storage probably we can reuse the volatile disk or files & kernels datastore. 

This will be implemented as a separated issue because of the implications and side effects on storage drivers.

In the meantime, note that this file is copied into the VM directory so it is consider during the VM life-cycle (i.e. migrations, undeploy/resume operations...).

**Comment by atodorov-storpool** [comment]
> In the meantime, note that this file is copied into the VM directory so it is consider during the VM life-cycle (i.e. migrations, undeploy/resume operations...).

Yes, this is covering the case when there is a file but it working with block devices too and in that case a there is no file copy (there is dd to the block device instead)... If the System TM_MAD is used, then the storage driver will take care to prepare the NVRAM file - File or symlink to a block device. Also, with the current implementation, it is not possible to backup the VM because there is no handling of the NVRAM files the backup extension implemented in the ruby bindings :)

It is not a strictly critical issue because our current [implementation](https://github.com/OpenNebula/addon-storpool/blob/master/docs/uefi_boot.md) could continue to be used, but I prefer such tweaks to be as streamlined as possible. In our implementation, the NVRAM file is prepared in the contextualization script and the domain XML is tweaked with a [deploy-tweak](https://github.com/OpenNebula/addon-storpool/blob/master/docs/deploy_tweaks.md). Migrate, undeploy, etc cases are handled in the relevant TM_MAD scripts too. The setup works with all recent opennebula versions currently in production, btw :)

**Comment by rsmontero** [comment]
Yes but note that for 6.2 it is always a file as it is copied from a file. Also note it is same case as with the checkpoint file.

I understand the idea, just wanted to point out, that any change on the NVRAM will be preserved if you migrate or undeploy the VM, for the current implementation.

**Comment by atodorov-storpool** [comment]
Yes, noted. It is indeed the same case with a ready resolution but not accepted for a long time...  :-/

**Comment by rdiaz-on** [comment]
>Having some experience with UEFI boot with OIpenNebula and after looking at the code I would like to understand some of the insights backing the current implementation?
>
>   There is no option/definition of the secure attribute set for the loader element at 121b4b0#diff-4f32f4a94d596188452be83043bf54091d4ded4e0d9e4b8fde332e124a3fa9fbR807 . It is an assumption that all distributions had qemu-kvm builds with secure=on enabled by default?
>
>   There is no option to define the SMM feature (SMM is needed for the Secure Boot). So it is assumed that it is active and enabled by default?

Thanks for the heads up. We forgot to include the `secure` tag and enable SMM (both required for secure boot to work). So it was only working for UEFI no secure. This [commit](https://github.com/OpenNebula/one/commit/66ce6f3c986c0e2122989ce9519153ac8aec9745) solves the problem.
